src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.cs

```diff
@@ -2,47 +2,239 @@
+                                counts = newCounts;
+
+                                if (newCounts.numThreadsGoal > oldCounts.numThreadsGoal && ThreadPool.GetQueuedWorkItems().Any())
+                                {
+                                    WorkerThread.MaybeAddWorkingWorker();
+                                }
+                            }
+                            else
+                            {
+                                counts = oldCounts;
+                            }
+                        }
+                    }
+                    return true;
+                }
             }
-            else
+            finally
             {
-                s_minThreads = threads;
-                return true;
+                s_maxMinThreadLock.Release();
             }
         }
 
         public static int GetMinThreads() => s_minThreads;
 
-        public static bool SetMaxThreads(int threads)
+        public static bool SetMaxThreads(int maxThreads)
         {
-            if (threads < s_minThreads || threads == 0)
+            s_maxMinThreadLock.Acquire();
+            try
             {
-                return false;
+                if (maxThreads < s_minThreads || maxThreads == 0)
+                {
+                    return false;
+                }
+                else
+                {
+                    short threads = (short)Math.Min(maxThreads, MaxPossibleThreadCount);
+                    if (s_forcedMaxWorkerThreads == 0)
+                    {
+                        s_maxThreads = threads;
+
+                        ThreadCounts counts = ThreadCounts.VolatileReadCounts(ref s_separated.counts);
+                        while (counts.numThreadsGoal > s_maxThreads)
+                        {
+                            ThreadCounts newCounts = counts;
+                            newCounts.numThreadsGoal = s_maxThreads;
+
+                            ThreadCounts oldCounts = ThreadCounts.CompareExchangeCounts(ref s_separated.counts, newCounts, counts);
+                            if (oldCounts == counts)
+                            {
+                                counts = newCounts;
+                            }
+                            else
+                            {
+                                counts = oldCounts;
+                            }
+                        }
+                    }
+                    return true;
+                }
             }
-            else
+            finally
             {
-                s_maxThreads = threads;
-                return true;
+                s_maxMinThreadLock.Release();
             }
         }
 
         public static int GetMaxThreads() => s_maxThreads;
+
+        public static int GetAvailableThreads()
+        {
+            ThreadCounts counts = ThreadCounts.VolatileReadCounts(ref s_separated.counts);
+            return counts.numExistingThreads - counts.numProcessingWork;
```

Comment:
This should return s_maxThreads - numExistingThreads, or zero if that would be negative